### PR TITLE
Fix error handling.

### PR DESCRIPTION
**Pull request by claygoddard**
A few spots needed error handling. Fix these spots to
better detect unrecoverable, unreported error cases.

### CODE REVIEW

File: solrmonitor/zkwatcherman.go [hunk 1]
```diff
@@ -175,8 +178,13 @@ func (m *zkWatcherMan) monitorChildren(synch bool, path string, onchange func(ch
 			}
 			return nil
 		}
-		cont := onchange(children)
-		if cont {
+		cont, err := onchange(children)
+		if err != nil {
+			// We failed to call the onchange handler; add a task for the recovery thread to keep trying
+			m.logger.Printf("zkWatcherMan %s: error calling onchange: %s", path, err)
+			m.enqueueDeferredTask(deferredChildrenTask{path: path, onchange: onchange})
```
Comment (donjaime) [inline solrmonitor/zkwatcherman.go]: Could this get stuck in an error loop and spam the recovery thread? Should enqueue here check a max queue length and then bail with a error log if it's over?

File: solrmonitor/solrmonitor.go [hunk 2]
```diff
@@ -183,7 +191,14 @@ func (c *SolrMonitor) updateCollections(collections []string, isInit bool) {
 			wg.Add(1)
 			go func() {
 				defer wg.Done()
-				coll.start(isInit)
+				if err := coll.start(isInit); err != nil {
+					// Only possible to get an error here if isInit = true,
+					// and in this case we can propagate the error all the way back.
+					select {
+					case errChan <- err:
+					default:
```
Comment (donjaime) [inline solrmonitor/solrmonitor.go]: Only the first error will propagate an error back. Which is by design. But do we want to consider some sort of multi-error or maybe some sort of atomic counter to allow us to track how many collections failed to start?

File: solrmonitor/zkwatcherman.go (hunk 1 above)
Comment (claygoddard) [reply solrmonitor/zkwatcherman.go]: I don't think the queue would grow unbounded. If we reach this point, the original `getChildrenAndWatch` already succeeded and the corresponding `monitorChildren` returned no error meaning the deferred task would be removed from the queue. Further errors would result in a new deferred task, but no ongoing watch. Basically a given collection should either have an ongoing watch or a deferred task, but never both.

File: solrmonitor/solrmonitor.go (hunk 2 above)
Comment (claygoddard) [reply solrmonitor/solrmonitor.go]: Sure! I'll at least keep a count and output num errors as part of the final log too.

**Pull request merged**